interactions/api/gateway.py

```diff
@@ -1,430 +1,321 @@
         "session_id",
         "sequence",
-        "keep_alive",
-        "closed",
-        "http",
-        "options",
+        "ready",
     )
 
     def __init__(
         self,
+        token: str,
         intents: Intents,
-        session_id: Optional[int] = None,
-        sequence: Optional[int] = None,
+        session_id: Optional[int] = MISSING,
+        sequence: Optional[int] = MISSING,
     ) -> None:
         """
-        :param intents: The intents used for identifying the connection.
+        :param token: The token of the application for connecting to the Gateway.
+        :type token: str
+        :param intents: The Gateway intents of the application for event dispatch.
         :type intents: Intents
-        :param session_id?: The session ID if you're trying to resume a connection. Defaults to ``None``.
+        :param session_id?: The ID of the session if trying to reconnect. Defaults to ``None``.
         :type session_id: Optional[int]
-        :param sequence?: The sequence if you're trying to resume a connection. Defaults to ``None``.
+        :param sequence?: The identifier sequence if trying to reconnect. Defaults to ``None``.
         :type sequence: Optional[int]
         """
-        self.intents = intents
-        self.loop = get_event_loop()
-        self.dispatch = Listener()
-        self.session = None
-        self.session_id = session_id
-        self.sequence = sequence
-
-        self.keep_alive = None
-        self.closed = False
-        self.http = None
-        self.options: dict = {
+        try:
+            self._loop = get_event_loop() if version_info < (3, 10) else get_running_loop()
+        except RuntimeError:
+            self._loop = new_event_loop()
+        self._dispatch = Listener()
+        self._http = HTTPClient(token)
+        self._client = None
+        self._closed = False
+        self._options = {
             "max_msg_size": 1024 ** 2,
             "timeout": 60,
             "autoclose": False,
             "compress": 0,
         }
-
-    async def recv(self) -> Optional[Any]:
-        """Receives packets sent from the gateway."""
-        packet = await self.session.receive()
-        return (
-            loads(packet.data)
-            if packet and isinstance(packet.data, (bytearray, bytes, memoryview, str))
-            else None
+        self._intents = intents
+        self.__heartbeater: _Heartbeat = _Heartbeat(
+            loop=self._loop if version_info < (3, 10) else None
         )
+        self.__shard = None
+        self.__presence = None
+        self.__task = None
+        self.session_id = None if session_id is MISSING else session_id
+        self.sequence = None if sequence is MISSING else sequence
+        self.ready = None
+
+    @property
+    async def __heartbeat_manager(self) -> None:
+        """Manages the heartbeat loop."""
+        while True:
+            if self.__heartbeater.event.is_set():
+                await self.__heartbeat_packet
```

Comment:
What happens if you try to send a heartbeat packet when the WebSocket is reconnecting?

Reply:
Will have to check, but it's likely nothing. When a new connection is established, we force the heartbeat delay back to 0 again to avoid an unsynced heartbeater

Reply:
I am saying that this line of code can execute when the WebSocket is in the process of reconnecting. Either you accidentally try to send a packet to a closed WebSocket, one that's in-fact None (`self._client` is None), or one that has not yet received the HELLO event / finished connecting.

Reply:
The manager only happens after we've declared `_client` as a `ClientWebSocketResponse` in the with loop, so I believe it's impossible to have a `TypeError` race condition.

We can add an extra condition that only triggers this when we've ensured the Gateway have sent us a `HELLO` payload. This ties in pretty well to our establisher, however, as any issue whatsoever outside of this should raise an exception when a closing code is found.